File: 350. Intersection of Two Arrays II/_350.py
class Solution:
    def intersect(self, nums1, nums2):
        from collections import Counter
        res = list()
        mydict = Counter(nums1)
        for i in nums2:
            if mydict[i]:
                res.append(i)
                mydict[i]-=1
        return res

    # 这种双指针的思路，对于两个有序的列表是有不错的效果的
    def intersect1(self, nums1, nums2):
        nums1, nums2 = sorted(nums1), sorted(nums2)
        pt1 = pt2 = 0
        res = []
        while True:
            try:
                if nums1[pt1] > nums2[pt2]:
                    pt2 += 1
                elif nums1[pt1] < nums2[pt2]:
                    pt1 += 1
                else:
                    res.append(nums1[pt1])
                    pt1 += 1
                    pt2 += 1
            except IndexError:
                break
        return res

File: 350. Intersection of Two Arrays II/350.go
package leetcode

func intersect(nums1 []int, nums2 []int) []int {
	myDict := make(map[int]int, 0)
	res := make([]int, 0)
	for _, v := range nums1{
		if _, ok := myDict[v]; ok{
			myDict[v]++
		}else{
			myDict[v]=1
		}
	}

	// map取不到值时，用的是默认值，比如这里的int就默认是0了，所以下面的写法更简洁
	// for _, n := range nums1 {
	// 	myDict[n]++
	// }

	for _, v := range nums2{
		if dictv, ok := myDict[v]; ok&&dictv!=0{
			myDict[v]--
			res = append(res, v)
		}
	}
	return res
}